File: Topics/Graphs/eTheMazeLC.cpp
/*
    Problem Link: https://www.lintcode.com/problem/787/
    Topic: Graphs - BFS
    Solution Video Link:
    Time Complexity:
    Space Complexity:  
*/

#include <bits/stdc++.h>
using namespace std;
#define ull unsigned long long int


bool hasPath(vector<vector<int>> &maze, vector<int> &start, vector<int> &destination){

	int n = maze.size();
	int m = maze[0].size();

	queue<pair<int, int>> q;
	q.push({start[0], start[1]});

	int vis[n][m];
	memset(vis, 0, sizeof vis);

	vis[start[0]][start[1]] = 1;

	while(!q.empty()){

		int x = q.front().first;
		int y = q.front().second;
		q.pop();

		if(x == destination[0] && y == destination[1]) 
			return true;


		int dx[] = {-1, 0, 1, 0};
		int dy[] = {0, 1, 0, -1};

		for(int ind = 0; ind < 4; ind++){
			int newX = x;
			int newY = y;

			while(newX >= 0 && newX < n && newY >= 0 && newY < n && maze[newX][newY] == 0){
				newX += dx[ind];
				newY += dy[ind];
			}
			newX -= dx[ind];
			newY -= dy[ind];

			if(!vis[newX][newY]){
				vis[newX][newY] = 1;
				q.push({newX, newY});
			}

		}

	}
	return false;


}

int main(){
#ifndef ONLINE_JUDGE
    // for getting input from input.txt
    freopen("input.txt", "r", stdin);
    // for writing output to output.txt
    freopen("output.txt", "w", stdout);
#endif

    return 0;

}

File: Topics/Graphs/eTheMazeLC.java
/*
    Problem Link: https://www.lintcode.com/problem/787/
    Topic: Graphs - BFS
    Solution Video Link:
    Time Complexity:
    Space Complexity:  
*/

import java.io.*;
import java.util.*;

public class eTheMazeLC {

	public static boolean hasPath(int[][] maze, int[] start, int[] destination) {
        // write your code here
        int n = maze.length;
        int m = maze[0].length;
        boolean[][] visited = new boolean[n][m];
        int[] dx = new int[]{0, -1, 0, 1};
        int[] dy = new int[]{1, 0, -1, 0};
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(start);
        visited[start[0]][start[1]] = true;
        while (!queue.isEmpty()) {
            int[] cur = queue.poll();
            if (cur[0] == destination[0] && cur[1] == destination[1]) {
                return true;
            }
            for (int direction = 0; direction < 4; direction++) {
                int nx = cur[0], ny = cur[1];
                while (nx >= 0 && nx < n && ny >= 0 && ny < m && maze[nx][ny] == 0) {
                    nx += dx[direction];
                    ny += dy[direction];
                }
                nx -= dx[direction];
                ny -= dy[direction];
                if (!visited[nx][ny]) {
                    visited[nx][ny] = true;
                    queue.offer(new int[]{nx, ny});
                }
            }
        }
        return false;
    }



	public static void main (String[] args) {

		try {
			System.setIn(new FileInputStream("input.txt"));
			System.setOut(new PrintStream(new FileOutputStream("output.txt")));
		} catch (Exception e) {
			System.err.println("Error");
		}
	    
	    //taking input using Scanner class
		Scanner sc=new Scanner(System.in);
		
		//taking total testcases
		int T=sc.nextInt();
		while(T-->0)
		{
		    
		}
		
	}
}